### PR TITLE
Gradient matching (Witches' Brew) for Tensorflow

### LINKED ISSUE #1386: Implement Witches’ brew  poisoning attack
**Is your feature request related to a problem? Please describe.**
Implement poisoning attack: 
`Witches' Brew: Industrial Scale Data Poisoning via Gradient Matching` 
https://arxiv.org/abs/2009.02276 

**Describe the solution you'd like**

- Jupyter notebook illustrating the use of the attack 
- Example folder 
- Implementation for Pytorch and Keras framework.

**Describe alternatives you've considered**
NA

**Additional context**
NA

### PR DESCRIPTION
**Pull request by TS-Lee**
# Description

This branch adds Gradient Matching (Witches' Brew) poisoning attack. This attack computes the clean label noise by setting the training gradient of poisoned samples towards the poisoning target class. This PR is for a TF version. The current commit is work in progress, and the follow up will come with a test and an example notebook.

Fixes #1386 

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS: Ubuntu 18.02
- Python version: 3.8
- ART version or commit number: dev_1.9.0
- TensorFlow / Keras / PyTorch / MXNet version: Tensorflow 2.4.1.

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/poisoning/gradient_matching_attack.py [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements Gradient Matching clean-label attacks (a.k.a. Witches' Brew) on Neural Networks.
+| Paper link: https://arxiv.org/abs/2009.02276
```
Comment (beat-buesser) [inline art/attacks/poisoning/gradient_matching_attack.py:20]:
```suggestion
This module implements Gradient Matching clean-label attacks (a.k.a. Witches' Brew) on Neural Networks.

| Paper link: https://arxiv.org/abs/2009.02276
```

File: art/attacks/poisoning/gradient_matching_attack.py [hunk 2]
```diff
@@ -0,0 +1,575 @@
...
+from art.estimators.classification.classifier import ClassifierMixin
+
+if TYPE_CHECKING:
+    # pylint: disable=C0412
+    from art.utils import CLASSIFIER_NEURALNETWORK_TYPE
+
+logger = logging.getLogger(__name__)
+
+
+class GradientMatchingAttack(Attack):
```
Comment (beat-buesser) [inline art/attacks/poisoning/gradient_matching_attack.py:41]:
```suggestion
class GradientMatchingAttack(PoisoningAttackWhiteBox):
```

File: art/attacks/poisoning/gradient_matching_attack.py [hunk 3]
```diff
@@ -0,0 +1,575 @@
...
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+from typing import Any, Dict, Tuple, TYPE_CHECKING, List
+
+import numpy as np
+from tqdm.auto import trange, tqdm
+
+from art.attacks.attack import Attack
```
Comment (beat-buesser) [inline art/attacks/poisoning/gradient_matching_attack.py:30]:
```suggestion
from art.attacks.attack import PoisoningAttackWhiteBox
```

File: notebooks/poisoning_attack_witches_brew.ipynb [hunk 4]
```diff
@@ -0,0 +1,1098 @@
+{
```
Comment (beat-buesser) [inline notebooks/poisoning_attack_witches_brew.ipynb:1]: Would it be possible to suppress the warnings in cell 3?

File: notebooks/poisoning_attack_witches_brew.ipynb [hunk 5]
```diff
@@ -0,0 +1,1098 @@
...
+ "cells": [
```
Comment (beat-buesser) [inline notebooks/poisoning_attack_witches_brew.ipynb:2]: Can you change the "TF" in the title to "TensorFlow v2"?

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @TS-Lee Thank you very much for your pull request! It looks good to me. I have left a few comments for small improvements.

File: art/attacks/poisoning/gradient_matching_attack.py (hunk 2 above)
Comment (TS-Lee) [reply art/attacks/poisoning/gradient_matching_attack.py:41]: I couldn't do this initially because of their different interface.
PoisoningAttackWhiteBox takes samples to poison and their target classes in poison method.
However, Gradient Matching attack needs (x_trigger, y_trigger) and (x_poison, y_poison). x_poison is samples to be manipulated, but the output changes for x_trigger instead, not x_poison. This can be confusing if we use the same interface, and Gradient Matching cannot replace other attacks implementing PoisoningAttackWhiteBox.

File: art/attacks/poisoning/gradient_matching_attack.py (hunk 3 above)
Comment (TS-Lee) [reply art/attacks/poisoning/gradient_matching_attack.py:31]: Same as above.
